### PR TITLE
Vanilla Crash & Furnace Speedup

### PR DESCRIPTION
**Pull request by mitchej123**
1) Backports a fix for URL parsing from https://github.com/MinecraftForge/MinecraftForge/issues/1712
2) Makes FurnaceRecipe significantly faster by using a HashMap w/ ItemStackHashingStrategy

### CODE REVIEW

File: src/main/java/com/mitchej123/hodgepodge/core/util/ItemStackHashingStrategy.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ *  - Taken from https://github.com/hilburn/AdvancedSystemsManager/blob/master/src/main/java/advancedsystemsmanager/flow/execution/buffers/maps/ItemStackHashingStrategy.java
+ *    under the the DBaJ (Don't Be a Jerk) non-commercial care-free license. 
+ *  (c) hilburn
+ */
+public class ItemStackHashingStrategy implements HashingStrategy<ItemStack> {
+    public static final ItemStackHashingStrategy INSTANCE = new ItemStackHashingStrategy();
+    
+    @Override
+    public int computeHashCode(ItemStack stack) {
+        return stack.getItem().hashCode() ^ stack.getItemDamage() ^ (stack.hasTagCompound() ? stack.stackTagCompound.hashCode() : 0);
```
Comment (repo-alt) [inline src/main/java/com/mitchej123/hodgepodge/core/util/ItemStackHashingStrategy.java:17]: Wouldn't it give better distribution if you e.g. multiply item damage by some large enough prime here? It's not that expensive

File: src/main/java/com/mitchej123/hodgepodge/asm/ThermosFurnaceSledgeHammer.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    private static final Logger LOGGER = LogManager.getLogger("ThermosFurnaceSledgeHammer");
+    @Override
+    public byte[] transform(String name, String transformedName, byte[] basicClass) {
+        if (HodgepodgeMixinPlugin.config.thermosCraftServerClass.equals(transformedName)) {
+            LOGGER.info("Patching Thermos or derivative to not break our furnace fix");
+            final ClassReader cr = new ClassReader(basicClass);
+            final ClassWriter cw = new ClassWriter(0);
+
+            final ClassNode cn = new ClassNode(ASM5);
+            cr.accept(cn, ClassReader.EXPAND_FRAMES);
```
Comment (Glease) [inline src/main/java/com/mitchej123/hodgepodge/asm/ThermosFurnaceSledgeHammer.java:28]: No need to expand frames.

File: src/main/java/com/mitchej123/hodgepodge/asm/ThermosFurnaceSledgeHammer.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            for (MethodNode m : cn.methods) {
+                if ("resetRecipes".equals(m.name)) {
+                    LOGGER.info("Taking a sledgehammer to CraftServer.resetRecipes()");
+                    //Replace the body with a RETURN opcode
+                    InsnList insnList = new InsnList();
+                    insnList.add(new InsnNode(Opcodes.RETURN));
+                    m.instructions = insnList;
```
Comment (Glease) [inline src/main/java/com/mitchej123/hodgepodge/asm/ThermosFurnaceSledgeHammer.java:35]: Probably should change m.maxStack to 0 as well.

**Review by Glease — COMMENTED** [review]
Beyond these, change the hashing function and we should be good to merge

**Review by Dream-Master — APPROVED** [review]

**Pull request merged**